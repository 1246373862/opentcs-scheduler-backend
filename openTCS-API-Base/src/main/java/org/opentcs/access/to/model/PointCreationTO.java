/**
 * Copyright (c) The openTCS Authors.
 *
 * This program is free software and subject to the MIT license. (For details,
 * see the licensing information (LICENSE.txt) you should have received with
 * this copy of the software.)
 */
package org.opentcs.access.to.model;

import java.io.Serializable;
import java.util.Map;
import static java.util.Objects.requireNonNull;
import javax.annotation.Nonnull;
import org.opentcs.access.to.CreationTO;
import org.opentcs.data.model.Couple;
import org.opentcs.data.model.Envelope;
import org.opentcs.data.model.Point;
import org.opentcs.data.model.Pose;
import org.opentcs.data.model.Triple;
import org.opentcs.util.annotations.ScheduledApiChange;

/**
 * A transfer object describing a point in the plant model.
 */
public class PointCreationTO
    extends CreationTO
    implements Serializable {

  /**
   * The pose of the vehicle at this point.
   */
  private final Pose pose;
  /**
   * This point's type.
   */
  @Nonnull
  private final Point.Type type;
  /**
   * A map of envelope keys to envelopes that vehicles located at this point may occupy.
   */
  private final Map<String, Envelope> vehicleEnvelopes;
  /**
   * The information regarding the grahical representation of this point.
   */
  private final Layout layout;

  /**
   * Creates a new instance.
   *
   * @param name The name of this point.
   */
  public PointCreationTO(@Nonnull String name) {
    super(name);
    this.pose = new Pose(new Triple(0, 0, 0), Double.NaN);
    this.type = Point.Type.HALT_POSITION;
    this.vehicleEnvelopes = Map.of();
    this.layout = new Layout();
  }

  private PointCreationTO(@Nonnull String name,
                          @Nonnull Map<String, String> properties,
                          @Nonnull Pose pose,
                          @Nonnull Point.Type type,
                          @Nonnull Map<String, Envelope> vehicleEnvelopes,
                          @Nonnull Layout layout) {
    super(name, properties);
    this.pose = requireNonNull(pose, "pose");
    this.type = requireNonNull(type, "type");
    this.vehicleEnvelopes = requireNonNull(vehicleEnvelopes, "vehicleEnvelopes");
    this.layout = requireNonNull(layout, "layout");
  }

  /**
   * Creates a copy of this object with the given name.
   *
   * @param name The new name.
   * @return A copy of this object, differing in the given name.
   */
  @Override
  public PointCreationTO withName(@Nonnull String name) {
    return new PointCreationTO(name,
                               getModifiableProperties(),
                               pose,
                               type,
                               vehicleEnvelopes,
                               layout);
  }

  /**
   * Returns the pose of the vehicle at this point.
   *
   * @return The pose of the vehicle at this point.
   */
  @Nonnull
  public Pose getPose() {
    return pose;
  }

  /**
   * Creates a copy of this object with the given pose.
   *
   * @param pose The new pose.
   * @return A copy of this object, differing in the given position.
   */
  public PointCreationTO withPose(@Nonnull Pose pose) {
    return new PointCreationTO(getName(),
                               getModifiableProperties(),
                               pose,
                               type,
                               vehicleEnvelopes,
                               layout);
  }

  /**
   * Returns the position of this point (in mm).
   *
   * @return The position of this point (in mm).
   * @deprecated Use {@link #getPose()} instead.
   */
  @Deprecated
  @ScheduledApiChange(when = "6.0", details = "Will be removed.")
  @Nonnull
  public Triple getPosition() {
    return pose.getPosition();
  }

  /**
   * Creates a copy of this object with the given position (in mm).
   *
   * @param position The new position.
   * @return A copy of this object, differing in the given position.
   * @deprecated Use {@link #withPose(org.opentcs.data.model.Pose)} instead.
   */
  @Deprecated
  @ScheduledApiChange(when = "6.0", details = "Will be removed.")
  public PointCreationTO withPosition(@Nonnull Triple position) {
    return new PointCreationTO(getName(),
                               getModifiableProperties(),
                               pose.withPosition(position),
                               type,
                               vehicleEnvelopes,
                               layout);
  }

  /**
   * Returns a vehicle's orientation angle at this position.
   * (-360..360, or {@code Double.NaN}, if an orientation angle is not specified for this point.)
   *
   * @return The vehicle's orientation angle when it's at this position.
   * @deprecated Use {@link #getPose()} instead.
   */
  @Deprecated
  @ScheduledApiChange(when = "6.0", details = "Will be removed.")
  public double getVehicleOrientationAngle() {
    return pose.getOrientationAngle();
  }

  /**
   * Creates a copy of this object with the vehicle's (assumed) orientation angle when it's at this
   * position.
   * Allowed value range: [-360..360], or {@code Double.NaN} to indicate that there is no specific
   * orientation angle for this point.
   *
   * @param vehicleOrientationAngle The new angle.
   * @return A copy of this object, differing in the given angle.
   * @deprecated Use {@link #withPose(org.opentcs.data.model.Pose)} instead.
   */
  @Deprecated
  @ScheduledApiChange(when = "6.0", details = "Will be removed.")
  public PointCreationTO withVehicleOrientationAngle(double vehicleOrientationAngle) {
    return new PointCreationTO(getName(),
                               getModifiableProperties(),
                               pose.withOrientationAngle(vehicleOrientationAngle),
                               type,
                               vehicleEnvelopes,
                               layout);
  }

  /**
   * Returns the type of this point.
   *
   * @return The type of this point.
   */
  @Nonnull
  public Point.Type getType() {
    return type;
  }

  /**
   * Creates a copy of this object with the given type.
   *
   * @param type The new type.
   * @return A copy of this object, differing in the given type.
   */
  public PointCreationTO withType(@Nonnull Point.Type type) {
    return new PointCreationTO(getName(),
                               getProperties(),
                               pose,
                               type,
                               vehicleEnvelopes,
                               layout);
  }

  /**
   * Creates a copy of this object with the given properties.
   *
   * @param properties The new properties.
   * @return A copy of this object, differing in the given properties.
   */
  @Override
  public PointCreationTO withProperties(@Nonnull Map<String, String> properties) {
    return new PointCreationTO(getName(),
                               properties,
                               pose,
                               type,
                               vehicleEnvelopes,
                               layout);
  }

  /**
   * Creates a copy of this object and adds the given property.
   * If value == null, then the key-value pair is removed from the properties.
   *
   * @param key the key.
   * @param value the value
   * @return A copy of this object that either
   * includes the given entry in it's current properties, if value != null or
   * excludes the entry otherwise.
   */
  @Override
  public PointCreationTO withProperty(@Nonnull String key, @Nonnull String value) {
    return new PointCreationTO(getName(),
                               propertiesWith(key, value),
                               pose,
                               type,
                               vehicleEnvelopes,
                               layout);
  }

  /**
   * Returns a map of envelope keys to envelopes that vehicles located at this point may occupy.
   *
   * @return A map of envelope keys to envelopes that vehicles located at this point may occupy.
   */
  public Map<String, Envelope> getVehicleEnvelopes() {
    return vehicleEnvelopes;
  }

  /**
   * Creates a copy of this object, with the given vehicle envelopes.
   *
   * @param vehicleEnvelopes The value to be set in the copy.
   * @return A copy of this object, differing in the given value.
   */
  public PointCreationTO withVehicleEnvelopes(@Nonnull Map<String, Envelope> vehicleEnvelopes) {
    return new PointCreationTO(getName(),
                               getModifiableProperties(),
                               pose,
                               type,
                               vehicleEnvelopes,
                               layout);
  }

  /**
   * Returns the information regarding the grahical representation of this point.
   *
   * @return The information regarding the grahical representation of this point.
   */
  public Layout getLayout() {
    return layout;
  }

  /**
   * Creates a copy of this object, with the given layout.
   *
   * @param layout The value to be set in the copy.
   * @return A copy of this object, differing in the given value.
   */
  public PointCreationTO withLayout(Layout layout) {
    return new PointCreationTO(getName(),
                               getModifiableProperties(),
                               pose,
                               type,
                               vehicleEnvelopes,
                               layout);
  }

  @Override
  public String toString() {
    return "PointCreationTO{"
        + "name=" + getName()
        + ", pose=" + pose
        + ", type=" + type
        + ", vehicleEnvelopes=" + vehicleEnvelopes
        + ", layout=" + layout
        + ", properties=" + getProperties()
        + '}';
  }

  /**
   * Contains information regarding the grahical representation of a point.
   */
  public static class Layout
      implements Serializable {

    /**
     * The coordinates at which the point is to be drawn (in mm).
     */
    private final Couple position;
    /**
     * The offset of the label's position to the point's position (in lu).
     */
    private final Couple labelOffset;
    /**
     * The ID of the layer on which the point is to be drawn.
     */
    private final int layerId;

    /**
     * Creates a new instance.
     */
    public Layout() {
      this(new Couple(0, 0), new Couple(0, 0), 0);
    }

    /**
     * Creates a new instance.
     *
     * @param position The coordinates at which the point is to be drawn (in mm).
     * @param labelOffset The offset of the label's position to the point's position (in lu).
     * @param layerId The ID of the layer on which the point is to be drawn.
     */
    public Layout(Couple position,
                  Couple labelOffset,
                  int layerId) {
      this.position = requireNonNull(position, "position");
      this.labelOffset = requireNonNull(labelOffset, "labelOffset");
      this.layerId = layerId;
    }

    /**
     * Returns the coordinates at which the point is to be drawn (in mm).
     *
     * @return The coordinates at which the point is to be drawn (in mm).
     */
    public Couple getPosition() {
      return position;
    }

    /**
     * Creates a copy of this object, with the given position.
     *
     * @param position The value to be set in the copy.
     * @return A copy of this object, differing in the given value.
     */
    public Layout withPosition(Couple position) {
      return new Layout(position,
                        labelOffset,
                        layerId);
    }

    /**
     * Returns the offset of the label's position to the point's position (in lu).
     *
     * @return The offset of the label's position to the point's position (in lu).
     */
    public Couple getLabelOffset() {
      return labelOffset;
    }

    /**
     * Creates a copy of this object, with the given X label offset.
     *
     * @param labelOffset The value to be set in the copy.
     * @return A copy of this object, differing in the given value.
     */
    public Layout withLabelOffset(Couple labelOffset) {
      return new Layout(position,
                        labelOffset,
                        layerId);
    }

    /**
     * Returns the ID of the layer on which the point is to be drawn.
     *
     * @return The layer ID.
     */
    public int getLayerId() {
      return layerId;
    }

    /**
     * Creates a copy of this object, with the given layer ID.
     *
     * @param layerId The value to be set in the copy.
     * @return A copy of this object, differing in the given value.
     */
    public Layout withLayerId(int layerId) {
      return new Layout(position,
                        labelOffset,
                        layerId);
    }

    @Override
    public String toString() {
      return "Layout{"
          + "position=" + position
          + ", labelOffset=" + labelOffset
          + ", layerId=" + layerId
          + '}';
    }
  }
}
