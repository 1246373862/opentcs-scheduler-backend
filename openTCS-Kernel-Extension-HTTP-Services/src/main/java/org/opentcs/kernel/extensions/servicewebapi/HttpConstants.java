/**
 * Copyright (c) The openTCS Authors.
 *
 * This program is free software and subject to the MIT license. (For details,
 * see the licensing information (LICENSE.txt) you should have received with
 * this copy of the software.)
 */
package org.opentcs.kernel.extensions.servicewebapi;

/**
 * Defines some HTTP-related constants.
 */
public class HttpConstants {

  /**
   * Name of the header that is expected to contain the API access keys.
   */
  public static final String HEADER_NAME_ACCESS_KEY = "X-Api-Access-Key";
  /**
   * Content type for plain text.
   */
  public static final String CONTENT_TYPE_TEXT_PLAIN_UTF8 = "text/plain; charset=utf-8";
  /**
   * Content type for JSON structures.
   */
  public static final String CONTENT_TYPE_APPLICATION_JSON_UTF8 = "application/json; charset=utf-8";

  /**
   * Prevents instantiation.
   */
  private HttpConstants() {
  }
}
