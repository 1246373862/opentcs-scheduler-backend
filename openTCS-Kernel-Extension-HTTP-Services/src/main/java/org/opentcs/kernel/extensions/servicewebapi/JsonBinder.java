/**
 * Copyright (c) The openTCS Authors.
 *
 * This program is free software and subject to the MIT license. (For details,
 * see the licensing information (LICENSE.txt) you should have received with
 * this copy of the software.)
 */
package org.opentcs.kernel.extensions.servicewebapi;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
import java.io.IOException;

/**
 * Binds JSON strings to objects and vice versa.
 */
public class JsonBinder {

  /**
   * Maps between objects and their JSON representations.
   */
  private final ObjectMapper objectMapper
      = new ObjectMapper()
          .registerModule(new JavaTimeModule())
          .disable(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS);

  /**
   * Creates a new instance.
   */
  public JsonBinder() {
  }

  /**
   * Maps the given JSON string to an object.
   *
   * @param <T> The type of object to map to.
   * @param jsonString The JSON string.
   * @param clazz The type of object to map to.
   * @return The object created from the JSON string.
   * @throws IllegalArgumentException In case there was a problem mapping the given object from
   * JSON.
   * (An IllegalArgumentException is mapped to HTTP status code 400, indicating a client error.)
   */
  public <T> T fromJson(String jsonString, Class<T> clazz)
      throws IllegalArgumentException {
    try {
      return objectMapper.readValue(jsonString, clazz);
    }
    catch (IOException exc) {
      throw new IllegalArgumentException("Could not parse JSON input", exc);
    }
  }

  /**
   * Maps the given object to a JSON string.
   *
   * @param object The object to be mapped.
   * @return The JSON string representation of the object.
   * @throws IllegalStateException In case there was a problem mapping the given object to JSON.
   * (An IllegalStateException is mapped to HTTP status code 500, indicating an internal error.)
   */
  public String toJson(Object object)
      throws IllegalStateException {
    try {
      return objectMapper
          .writerWithDefaultPrettyPrinter()
          .writeValueAsString(object);
    }
    catch (JsonProcessingException exc) {
      throw new IllegalStateException("Could not produce JSON output", exc);
    }
  }

  /**
   * Maps the given throwable to a JSON string.
   *
   * @param t The throwable to be mapped.
   * @return A JSON string for the given throwable, consisting of a single-element array containing
   * the throwable's message.
   * @throws IllegalStateException In case there was a problem mapping the given object to JSON.
   * (An IllegalStateException is mapped to HTTP status code 500, indicating an internal error.)
   */
  public String toJson(Throwable t)
      throws IllegalStateException {
    try {
      return objectMapper
          .writerWithDefaultPrettyPrinter()
          .writeValueAsString(objectMapper.createArrayNode().add(t.getMessage()));
    }
    catch (JsonProcessingException exc) {
      throw new IllegalStateException("Could not produce JSON output", exc);
    }
  }

}
