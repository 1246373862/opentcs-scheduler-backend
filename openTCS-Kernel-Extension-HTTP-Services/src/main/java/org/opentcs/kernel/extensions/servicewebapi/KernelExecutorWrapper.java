/**
 * Copyright (c) The openTCS Authors.
 *
 * This program is free software and subject to the MIT license. (For details,
 * see the licensing information (LICENSE.txt) you should have received with
 * this copy of the software.)
 */
package org.opentcs.kernel.extensions.servicewebapi;

import static java.util.Objects.requireNonNull;
import java.util.concurrent.Callable;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import javax.inject.Inject;
import org.opentcs.access.KernelRuntimeException;
import org.opentcs.customizations.kernel.KernelExecutor;

/**
 * Calls callables/runnables via the kernel executor and waits for the outcome.
 */
public class KernelExecutorWrapper {

  private final ExecutorService kernelExecutor;

  /**
   * Creates a new instance.
   *
   * @param kernelExecutor The kernel executor.
   */
  @Inject
  public KernelExecutorWrapper(@KernelExecutor ExecutorService kernelExecutor) {
    this.kernelExecutor = requireNonNull(kernelExecutor, "kernelExecutor");
  }

  /**
   * Calls the given callable via the kernel executor and waits for the outcome.
   *
   * @param <T> The callable's return type.
   * @param callable The callable.
   * @return The result of the call.
   * @throws IllegalStateException In case the call via the kernel executor was unexpectedly
   * interrupted.
   * @throws RuntimeException In case an exception was thrown from the callable. If the exception
   * thrown is a {@code RuntimeException}, it is forwarded directly; if it is not a
   * {@code RuntimeException}, it is wrapped in a {@link KernelRuntimeException}.
   */
  public <T> T callAndWait(Callable<T> callable)
      throws IllegalStateException, RuntimeException {
    requireNonNull(callable, "callable");

    try {
      return kernelExecutor.submit(callable).get();
    }
    catch (InterruptedException exc) {
      throw new IllegalStateException("Unexpectedly interrupted");
    }
    catch (ExecutionException exc) {
      if (exc.getCause() instanceof RuntimeException) {
        throw (RuntimeException) exc.getCause();
      }
      throw new KernelRuntimeException(exc.getCause());
    }
  }

  /**
   * Calls the given runnable via the kernel executor and waits for the outcome.
   *
   * @param runnable The runnable.
   * @throws IllegalStateException In case the call via the kernel executor was unexpectedly
   * interrupted.
   * @throws RuntimeException In case an exception was thrown from the runnable. If the exception
   * thrown is a {@code RuntimeException}, it is forwarded directly; if it is not a
   * {@code RuntimeException}, it is wrapped in a {@link KernelRuntimeException}.
   */
  public void callAndWait(Runnable runnable)
      throws IllegalStateException, RuntimeException {
    requireNonNull(runnable, "runnable");

    callAndWait(Executors.callable(runnable));
  }
}
