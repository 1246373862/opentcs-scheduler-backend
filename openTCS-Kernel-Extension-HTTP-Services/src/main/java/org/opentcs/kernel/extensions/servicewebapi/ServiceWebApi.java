/**
 * Copyright (c) The openTCS Authors.
 *
 * This program is free software and subject to the MIT license. (For details,
 * see the licensing information (LICENSE.txt) you should have received with
 * this copy of the software.)
 */
package org.opentcs.kernel.extensions.servicewebapi;

import com.google.common.util.concurrent.Uninterruptibles;
import static java.util.Objects.requireNonNull;
import java.util.concurrent.TimeUnit;
import javax.inject.Inject;
import org.opentcs.access.KernelRuntimeException;
import org.opentcs.access.SslParameterSet;
import org.opentcs.components.kernel.KernelExtension;
import org.opentcs.data.ObjectExistsException;
import org.opentcs.data.ObjectUnknownException;
import org.opentcs.kernel.extensions.servicewebapi.v1.V1RequestHandler;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import spark.Service;

/**
 * Provides an HTTP interface for basic administration needs.
 */
public class ServiceWebApi
    implements KernelExtension {

  /**
   * This class's logger.
   */
  private static final Logger LOG = LoggerFactory.getLogger(ServiceWebApi.class);
  /**
   * The interface configuration.
   */
  private final ServiceWebApiConfiguration configuration;
  /**
   * Authenticates incoming requests.
   */
  private final Authenticator authenticator;
  /**
   * Handles requests for API version 1.
   */
  private final V1RequestHandler v1RequestHandler;
  /**
   * Binds JSON data to objects and vice versa.
   */
  private final JsonBinder jsonBinder;
  /**
   * The connection encryption configuration.
   */
  private final SslParameterSet sslParamSet;
  /**
   * The actual HTTP service.
   */
  private Service service;
  /**
   * Whether this kernel extension is initialized.
   */
  private boolean initialized;

  /**
   * Creates a new instance.
   *
   * @param configuration The interface configuration.
   * @param sslParamSet The SSL parameter set.
   * @param authenticator Authenticates incoming requests.
   * @param jsonBinder Binds JSON data to objects and vice versa.
   * @param v1RequestHandler Handles requests for API version 1.
   */
  @Inject
  public ServiceWebApi(ServiceWebApiConfiguration configuration,
                       SslParameterSet sslParamSet,
                       Authenticator authenticator,
                       JsonBinder jsonBinder,
                       V1RequestHandler v1RequestHandler) {
    this.configuration = requireNonNull(configuration, "configuration");
    this.sslParamSet = requireNonNull(sslParamSet, "sslParamSet");
    this.authenticator = requireNonNull(authenticator, "authenticator");
    this.jsonBinder = requireNonNull(jsonBinder, "jsonBinder");
    this.v1RequestHandler = requireNonNull(v1RequestHandler, "v1RequestHandler");
  }

  @Override
  public void initialize() {
    if (isInitialized()) {
      return;
    }

    v1RequestHandler.initialize();

    service = Service.ignite()
        .ipAddress(configuration.bindAddress())
        .port(configuration.bindPort());

    if (configuration.useSsl()) {
      service.secure(sslParamSet.getKeystoreFile().getAbsolutePath(),
                     sslParamSet.getKeystorePassword(),
                     null,
                     null);
    }
    else {
      LOG.warn("Encryption disabled, connections will not be secured!");
    }

    service.before((request, response) -> {
      if (!authenticator.isAuthenticated(request)) {
        // Delay the response a bit to slow down brute force attacks.
        Uninterruptibles.sleepUninterruptibly(2, TimeUnit.SECONDS);
        service.halt(403, "Not authenticated.");
      }

      // Add a CORS header to allow cross-origin requests from all hosts.
      // This also makes using the "try it out" buttons in the Swagger UI documentation possible.
      response.header("Access-Control-Allow-Origin", "*");
    });

    // Reflect that we allow cross-origin requests for any headers and methods.
    service.options(
        "/*",
        (request, response) -> {
          String accessControlRequestHeaders = request.headers("Access-Control-Request-Headers");
          if (accessControlRequestHeaders != null) {
            response.header("Access-Control-Allow-Headers", accessControlRequestHeaders);
          }

          String accessControlRequestMethod = request.headers("Access-Control-Request-Method");
          if (accessControlRequestMethod != null) {
            response.header("Access-Control-Allow-Methods", accessControlRequestMethod);
          }

          return "OK";
        });

    // Register routes for API versions here.
    service.path("/v1", () -> v1RequestHandler.addRoutes(service));

    service.exception(IllegalArgumentException.class, (exception, request, response) -> {
                    response.status(400);
                    response.type(HttpConstants.CONTENT_TYPE_APPLICATION_JSON_UTF8);
                    response.body(jsonBinder.toJson(exception));
                  });
    service.exception(ObjectUnknownException.class, (exception, request, response) -> {
                    response.status(404);
                    response.type(HttpConstants.CONTENT_TYPE_APPLICATION_JSON_UTF8);
                    response.body(jsonBinder.toJson(exception));
                  });
    service.exception(ObjectExistsException.class, (exception, request, response) -> {
                    response.status(409);
                    response.type(HttpConstants.CONTENT_TYPE_APPLICATION_JSON_UTF8);
                    response.body(jsonBinder.toJson(exception));
                  });
    service.exception(KernelRuntimeException.class, (exception, request, response) -> {
                    response.status(500);
                    response.type(HttpConstants.CONTENT_TYPE_APPLICATION_JSON_UTF8);
                    response.body(jsonBinder.toJson(exception));
                  });
    service.exception(IllegalStateException.class, (exception, request, response) -> {
                    response.status(500);
                    response.type(HttpConstants.CONTENT_TYPE_APPLICATION_JSON_UTF8);
                    response.body(jsonBinder.toJson(exception));
                  });

    initialized = true;
  }

  @Override
  public void terminate() {
    if (!isInitialized()) {
      return;
    }

    v1RequestHandler.terminate();
    service.stop();

    initialized = false;
  }

  @Override
  public boolean isInitialized() {
    return initialized;
  }

}
