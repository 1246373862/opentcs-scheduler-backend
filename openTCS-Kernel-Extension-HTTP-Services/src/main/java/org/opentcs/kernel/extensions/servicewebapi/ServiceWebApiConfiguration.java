/**
 * Copyright (c) The openTCS Authors.
 *
 * This program is free software and subject to the MIT license. (For details,
 * see the licensing information (LICENSE.txt) you should have received with
 * this copy of the software.)
 */
package org.opentcs.kernel.extensions.servicewebapi;

import org.opentcs.configuration.ConfigurationEntry;
import org.opentcs.configuration.ConfigurationPrefix;

/**
 * Configuration entries for the service web API.
 */
@ConfigurationPrefix(ServiceWebApiConfiguration.PREFIX)
public interface ServiceWebApiConfiguration {

  /**
   * The prefix for all configuration entries here.
   */
  String PREFIX = "servicewebapi";

  @ConfigurationEntry(
      type = "Boolean",
      description = "Whether to enable the interface.",
      changesApplied = ConfigurationEntry.ChangesApplied.ON_APPLICATION_START,
      orderKey = "0"
  )
  boolean enable();

  @ConfigurationEntry(
      type = "IP address",
      description = "Address to which to bind the HTTP server, e.g. 0.0.0.0 or 127.0.0.1.",
      changesApplied = ConfigurationEntry.ChangesApplied.ON_APPLICATION_START,
      orderKey = "1"
  )
  String bindAddress();

  @ConfigurationEntry(
      type = "Integer",
      description = "Port to which to bind the HTTP server.",
      changesApplied = ConfigurationEntry.ChangesApplied.ON_APPLICATION_START,
      orderKey = "2"
  )
  int bindPort();

  @ConfigurationEntry(
      type = "String",
      description = "Key allowing access to the API.",
      changesApplied = ConfigurationEntry.ChangesApplied.INSTANTLY,
      orderKey = "3"
  )
  String accessKey();

  @ConfigurationEntry(
      type = "Integer",
      description = "Maximum number of status events to be kept.",
      changesApplied = ConfigurationEntry.ChangesApplied.INSTANTLY,
      orderKey = "4"
  )
  int statusEventsCapacity();

  @ConfigurationEntry(
      type = "Boolean",
      description = "Whether to use SSL to encrypt connections.",
      changesApplied = ConfigurationEntry.ChangesApplied.ON_APPLICATION_START,
      orderKey = "5"
  )
  boolean useSsl();
}
