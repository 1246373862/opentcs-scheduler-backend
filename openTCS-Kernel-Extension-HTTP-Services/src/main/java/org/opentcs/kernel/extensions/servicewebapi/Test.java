package org.opentcs.kernel.extensions.servicewebapi;

import org.opentcs.components.kernel.KernelExtension;

public class Test implements KernelExtension {

  private volatile boolean enabled;

  @Override
  public void initialize() {
    if (enabled){
      return;
    }
    System.out.println("初始化~~~~~初始化~~~~~初始化~~~~~初始化~~~~~初始化~~~~~初始化~~~~~");
  }

  @Override
  public boolean isInitialized() {
    return enabled;
  }

  @Override
  public void terminate() {
    if (!enabled) {
      return;
    }
    enabled = false;
    System.out.println("结束！！");
  }
}
