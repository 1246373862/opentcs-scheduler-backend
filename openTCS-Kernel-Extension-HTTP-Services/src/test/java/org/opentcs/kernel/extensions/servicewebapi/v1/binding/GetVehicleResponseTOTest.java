/**
 * Copyright (c) The openTCS Authors.
 *
 * This program is free software and subject to the MIT license. (For details,
 * see the licensing information (LICENSE.txt) you should have received with
 * this copy of the software.)
 */
package org.opentcs.kernel.extensions.servicewebapi.v1.binding;

import java.util.List;
import java.util.Map;
import java.util.TreeMap;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.opentcs.data.model.Vehicle;
import org.opentcs.kernel.extensions.servicewebapi.JsonBinder;

/**
 * Unit tests for {@link GetVehicleResponseTO}.
 */
public class GetVehicleResponseTOTest {

  private JsonBinder jsonBinder;

  @BeforeEach
  public void setUp() {
    jsonBinder = new JsonBinder();
  }

  @Test
  public void jsonSample() {
    GetVehicleResponseTO to
        = new GetVehicleResponseTO()
            .setName("some-vehicle")
            .setProperties(
                new TreeMap<>(
                    Map.of(
                        "some-key", "some-value",
                        "some-other-key", "some-other-value"
                    )
                )
            )
            .setLength(1234)
            .setEnergyLevelGood(90)
            .setEnergyLevelCritical(30)
            .setEnergyLevel(48)
            .setIntegrationLevel(Vehicle.IntegrationLevel.TO_BE_UTILIZED)
            .setPaused(false)
            .setProcState(Vehicle.ProcState.PROCESSING_ORDER)
            .setTransportOrder("some-order")
            .setCurrentPosition("some-point")
            .setPrecisePosition(new GetVehicleResponseTO.PrecisePosition(1, 2, 3))
            .setState(Vehicle.State.EXECUTING)
            .setEnvelopeKey("envelopeType-01")
            .setAllocatedResources(
                List.of(
                    List.of("some-path", "some-point"),
                    List.of("some-other-path", "some-other-point")
                )
            )
            .setClaimedResources(
                List.of(
                    List.of("some-path", "some-point"),
                    List.of("some-other-path", "some-other-point")
                )
            )
            .setAllowedOrderTypes(
                List.of(
                    "OrderType001",
                    "OrderType002"
                )
            );

    Approvals.verify(jsonBinder.toJson(to));
  }

}
