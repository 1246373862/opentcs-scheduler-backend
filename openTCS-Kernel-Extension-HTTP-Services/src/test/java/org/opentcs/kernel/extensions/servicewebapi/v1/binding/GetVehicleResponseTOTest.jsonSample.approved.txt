{
  "name" : "some-vehicle",
  "properties" : {
    "some-key" : "some-value",
    "some-other-key" : "some-other-value"
  },
  "length" : 1234,
  "energyLevelGood" : 90,
  "energyLevelCritical" : 30,
  "energyLevel" : 48,
  "integrationLevel" : "TO_BE_UTILIZED",
  "paused" : false,
  "procState" : "PROCESSING_ORDER",
  "transportOrder" : "some-order",
  "currentPosition" : "some-point",
  "precisePosition" : {
    "x" : 1,
    "y" : 2,
    "z" : 3
  },
  "state" : "EXECUTING",
  "allocatedResources" : [ [ "some-path", "some-point" ], [ "some-other-path", "some-other-point" ] ],
  "claimedResources" : [ [ "some-path", "some-point" ], [ "some-other-path", "some-other-point" ] ],
  "allowedOrderTypes" : [ "OrderType001", "OrderType002" ],
  "envelopeKey" : "envelopeType-01"
}
